### ISSUE TITLE
Trying to get a certain format...need help

### ISSUE THREAD

**Issue by ndwyer-paa** [pmoule/go2hal#6]
`{
  "_embedded": {
    "filters": [
      {
        "name": "condition",
        "label": "Condition",
        "type": "select",
        "options": [
          {
            "label": "Broken Ankle",
            "value": "broken_ankle"
          }
        ]
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/episode/filters"
    }
  }
}`

I'm trying to achieve something like the above but the `options` section of the array of resources is throwing me of api wise. Right now I have filters as a ResourceRelation to the root node. `filter` is an array of hal.Resources which should contain an `options` array but not sure how to accomplish this since it seems like `options` should be a ResourceRelation. Help would be grateful.

**Comment by ndwyer-paa** [comment]
Right now what I'm getting is the following 

`{
    "_embedded": {
        "filters": [
            {
                "_embedded": {
                    "options": [
                        {
                            "label": "Broken Ankle",
                            "value": "broken_ankle"
                        }
                    ]
                },
                "label": "Condition",
                "name": "condition",
                "type": "select"
            }
        ]
    },
    "_links": {
        "self": {
            "href": "/api/episode/filters"
        }
    }
}`

which is close to what I want but I don't want the `_embedded` on the inner Resources. This should be an option as I thought that `_embedded` was optional?

**Comment by pmoule** [comment]
Hi,

I'm not quite sure how and where to start.
First, this example should lead to JSON as described in your first post.

``` func TestXXX(t *testing.T) {
	factory := NewResourceFactory(nil)

	self := "http://example.com/filters"
	root := factory.CreateRootResource(self)

	filtersRel := factory.CreateResourceLink("filters", "")
	filters := []Filter{Filter{ID: 1, Name: "condition", Options: []string{"one", "two"}}}

	var embeddedFilters []Resource

	for _, f := range filters {
		embSelf := fmt.Sprintf("http://example.com/filters/%d", f.ID)
		embFilter := factory.CreateEmbeddedResource(embSelf)
		embFilter.AddData(f)
		embeddedFilters = append(embeddedFilters, embFilter)
	}

	filtersRel.SetResources(embeddedFilters)
	root.AddResource(filtersRel)

```
In this case 'options' is a simple array. It depends on your API and what you decide to provide with a call to http://example.com/filters. All filter details or just some useful information like e.g. 'name'.
Following each embedded resource's self link may provide more resource information like e.g. 'options'.

In case you are new to HAL, start reading some introduction from the HAL inventor himself
http://stateless.co/hal_specification.html. Maybe, you already know this.

tl;dr but this seems to be a quite good introduction, too
https://dzone.com/articles/introduction-hypertext-0

In case the provided [README](https://github.com/pmoule/go2hal/blob/master/README.md) should be improved, feel free to make a proposal.

Best regards 
Patrick

**Comment by ndwyer-paa** [comment]
Understand....I can do an array of string elements currently but I need options to be an array of json objects and cannot figure out how to do that via the api since the json objects are basically resources.  Thanks so much for responding by the way.

**Comment by ndwyer-paa** [comment]
so options would be something like 
``` options: [ { "name": "something", "value": "someval"}, { "name": "something2", "value": "someval2"}]```

**Comment by pmoule** [comment]
I added some structs to extend the example.
```
type Option struct {
	Label string `json:"label"` //tagging required for mapping
	Value string `json:"value"` //tagging required for mapping
}

type Filter struct {
	ID      int //no tag, no mapping :)
	Name    string   `json:"name"` //tagging required for mapping
	Options []Option `json:"options"` //tagging required for mapping
}
...
option := Option{Label: "Label", Value: "value"}
filters := []Filter{Filter{ID: 1, Name: "condition", Options: []Option{option}}}
```
Only root resources and embedded resources are required to be created via API. All resource properties 
just have to be valid JSON. So, if you like to attach data like `[]Option` just do it. Just keep in mind to tag all fields you want to be mapped to created JSON.

**Comment by ndwyer-paa** [comment]
Thanks for the added struct examples. That helped me solve my problem. I'm new to go but have used HAL in other languages and I was thinking about it wrong. I should have let my structs do the heavy lifting. Instead I was ommiting `Options` from the struct and was doing `hal.NewResourceRelation("options")`. Thanks for the patience and the actual effort you put into answering my noobish question.

**Comment by pmoule** [comment]
:) you're welcome!
